File: app/src/main/java/com/noxvision/app/hunting/moon/MoonPhaseCalculator.kt
package com.noxvision.app.hunting.moon

import java.util.Calendar
import java.util.TimeZone
import kotlin.math.floor

object MoonPhaseCalculator {
    // Synodic month in days (average time between new moons)
    private const val SYNODIC_MONTH = 29.530588853

    // Reference new moon: January 6, 2000, 18:14 UTC
    private const val REFERENCE_NEW_MOON_JD = 2451550.1

    enum class MoonPhase(val germanName: String, val icon: String, val illumination: String) {
        NEW_MOON("Neumond", "\uD83C\uDF11", "0%"),
        WAXING_CRESCENT("Zunehmender Sichelmond", "\uD83C\uDF12", "1-49%"),
        FIRST_QUARTER("Erstes Viertel", "\uD83C\uDF13", "50%"),
        WAXING_GIBBOUS("Zunehmender Dreiviertelmond", "\uD83C\uDF14", "51-99%"),
        FULL_MOON("Vollmond", "\uD83C\uDF15", "100%"),
        WANING_GIBBOUS("Abnehmender Dreiviertelmond", "\uD83C\uDF16", "99-51%"),
        LAST_QUARTER("Letztes Viertel", "\uD83C\uDF17", "50%"),
        WANING_CRESCENT("Abnehmender Sichelmond", "\uD83C\uDF18", "49-1%")
    }

    data class MoonInfo(
        val phase: MoonPhase,
        val daysSinceNewMoon: Double,
        val daysUntilNextNewMoon: Double,
        val illuminationPercent: Double,
        val activityPrediction: WildlifeActivityPrediction
    )

    enum class WildlifeActivityPrediction(val germanText: String, val level: Int) {
        VERY_HIGH("Sehr hohe Aktivitaet", 5),
        HIGH("Hohe Aktivitaet", 4),
        MEDIUM("Mittlere Aktivitaet", 3),
        LOW("Geringe Aktivitaet", 2),
        VERY_LOW("Sehr geringe Aktivitaet", 1)
    }

    fun calculateMoonPhase(timestamp: Long = System.currentTimeMillis()): MoonInfo {
        val julianDate = toJulianDate(timestamp)
        val daysSinceRef = julianDate - REFERENCE_NEW_MOON_JD
        val lunarCycles = daysSinceRef / SYNODIC_MONTH
        val currentCycleProgress = lunarCycles - floor(lunarCycles)
        val daysSinceNewMoon = currentCycleProgress * SYNODIC_MONTH
        val daysUntilNextNewMoon = SYNODIC_MONTH - daysSinceNewMoon

        val phase = getPhaseFromCycleProgress(currentCycleProgress)
        val illumination = calculateIllumination(currentCycleProgress)
        val activity = predictWildlifeActivity(phase, illumination)

        return MoonInfo(
            phase = phase,
            daysSinceNewMoon = daysSinceNewMoon,
            daysUntilNextNewMoon = daysUntilNextNewMoon,
            illuminationPercent = illumination,
            activityPrediction = activity
        )
    }

    private fun toJulianDate(timestamp: Long): Double {
        val calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        calendar.timeInMillis = timestamp

        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH) + 1
        val day = calendar.get(Calendar.DAY_OF_MONTH)
        val hour = calendar.get(Calendar.HOUR_OF_DAY)
        val minute = calendar.get(Calendar.MINUTE)
        val second = calendar.get(Calendar.SECOND)

        val dayFraction = (hour + minute / 60.0 + second / 3600.0) / 24.0

        val a = (14 - month) / 12
        val y = year + 4800 - a
        val m = month + 12 * a - 3

        val jdn = day + (153 * m + 2) / 5 + 365 * y + y / 4 - y / 100 + y / 400 - 32045

        return jdn + dayFraction - 0.5
    }

    private fun getPhaseFromCycleProgress(progress: Double): MoonPhase {
        return when {
            progress < 0.0625 -> MoonPhase.NEW_MOON
            progress < 0.1875 -> MoonPhase.WAXING_CRESCENT
            progress < 0.3125 -> MoonPhase.FIRST_QUARTER
            progress < 0.4375 -> MoonPhase.WAXING_GIBBOUS
            progress < 0.5625 -> MoonPhase.FULL_MOON
            progress < 0.6875 -> MoonPhase.WANING_GIBBOUS
            progress < 0.8125 -> MoonPhase.LAST_QUARTER
            progress < 0.9375 -> MoonPhase.WANING_CRESCENT
            else -> MoonPhase.NEW_MOON
        }
    }

    private fun calculateIllumination(progress: Double): Double {
        // Using cosine function for illumination approximation
        // At new moon (progress = 0), illumination = 0
        // At full moon (progress = 0.5), illumination = 100
        val angle = progress * 2 * Math.PI
        return ((1 - kotlin.math.cos(angle)) / 2) * 100
    }

    private fun predictWildlifeActivity(phase: MoonPhase, illumination: Double): WildlifeActivityPrediction {
        // Wildlife activity tends to be higher during bright moonlit nights
        // and around the full moon period
        return when (phase) {
            MoonPhase.FULL_MOON -> WildlifeActivityPrediction.VERY_HIGH
            MoonPhase.WAXING_GIBBOUS, MoonPhase.WANING_GIBBOUS -> WildlifeActivityPrediction.HIGH
            MoonPhase.FIRST_QUARTER, MoonPhase.LAST_QUARTER -> WildlifeActivityPrediction.MEDIUM
            MoonPhase.WAXING_CRESCENT, MoonPhase.WANING_CRESCENT -> WildlifeActivityPrediction.LOW
            MoonPhase.NEW_MOON -> WildlifeActivityPrediction.VERY_LOW
        }
    }

    fun getNextFullMoon(fromTimestamp: Long = System.currentTimeMillis()): Long {
        val info = calculateMoonPhase(fromTimestamp)
        val daysUntilFull = when {
            info.daysSinceNewMoon < SYNODIC_MONTH / 2 -> (SYNODIC_MONTH / 2) - info.daysSinceNewMoon
            else -> SYNODIC_MONTH - info.daysSinceNewMoon + (SYNODIC_MONTH / 2)
        }
        return fromTimestamp + (daysUntilFull * 24 * 60 * 60 * 1000).toLong()
    }

    fun getNextNewMoon(fromTimestamp: Long = System.currentTimeMillis()): Long {
        val info = calculateMoonPhase(fromTimestamp)
        return fromTimestamp + (info.daysUntilNextNewMoon * 24 * 60 * 60 * 1000).toLong()
    }
}

File: NoxVision-iOS/NoxVision/Hunting/Moon/MoonPhaseCalculator.swift
import Foundation

struct MoonPhaseCalculator {
    static let synodicMonth = 29.530588853
    static let referenceNewMoonJD = 2451550.2597222

    enum MoonPhase: String, CaseIterable {
        case newMoon
        case waxingCrescent
        case firstQuarter
        case waxingGibbous
        case fullMoon
        case waningGibbous
        case lastQuarter
        case waningCrescent

        var germanName: String {
            switch self {
            case .newMoon: return "Neumond"
            case .waxingCrescent: return "Zunehmender Sichelmond"
            case .firstQuarter: return "Erstes Viertel"
            case .waxingGibbous: return "Zunehmender Dreiviertelmond"
            case .fullMoon: return "Vollmond"
            case .waningGibbous: return "Abnehmender Dreiviertelmond"
            case .lastQuarter: return "Letztes Viertel"
            case .waningCrescent: return "Abnehmender Sichelmond"
            }
        }

        var icon: String {
            switch self {
            case .newMoon: return "\u{1F311}"
            case .waxingCrescent: return "\u{1F312}"
            case .firstQuarter: return "\u{1F313}"
            case .waxingGibbous: return "\u{1F314}"
            case .fullMoon: return "\u{1F315}"
            case .waningGibbous: return "\u{1F316}"
            case .lastQuarter: return "\u{1F317}"
            case .waningCrescent: return "\u{1F318}"
            }
        }

        var illumination: String {
            switch self {
            case .newMoon: return "0%"
            case .waxingCrescent: return "1-49%"
            case .firstQuarter: return "50%"
            case .waxingGibbous: return "51-99%"
            case .fullMoon: return "100%"
            case .waningGibbous: return "99-51%"
            case .lastQuarter: return "50%"
            case .waningCrescent: return "49-1%"
            }
        }
    }

    struct MoonInfo {
        let phase: MoonPhase
        let daysSinceNewMoon: Double
        let daysUntilNextNewMoon: Double
        let illuminationPercent: Double
        let activityPrediction: WildlifeActivityPrediction
    }

    enum WildlifeActivityPrediction: Int {
        case veryLow = 1
        case low = 2
        case medium = 3
        case high = 4
        case veryHigh = 5

        var germanText: String {
            switch self {
            case .veryHigh: return "Sehr hohe Aktivitaet"
            case .high: return "Hohe Aktivitaet"
            case .medium: return "Mittlere Aktivitaet"
            case .low: return "Geringe Aktivitaet"
            case .veryLow: return "Sehr geringe Aktivitaet"
            }
        }
    }

    static func calculateMoonPhase(timestamp: Date = Date()) -> MoonInfo {
        let julianDate = toJulianDate(timestamp)
        let daysSinceRef = julianDate - referenceNewMoonJD
        let lunarCycles = daysSinceRef / synodicMonth
        let currentCycleProgress = lunarCycles - floor(lunarCycles)
        let daysSinceNewMoon = currentCycleProgress * synodicMonth
        let daysUntilNextNewMoon = synodicMonth - daysSinceNewMoon

        let phase = getPhaseFromCycleProgress(currentCycleProgress)
        let illumination = calculateIllumination(currentCycleProgress)
        let activity = predictWildlifeActivity(phase)

        return MoonInfo(
            phase: phase,
            daysSinceNewMoon: daysSinceNewMoon,
            daysUntilNextNewMoon: daysUntilNextNewMoon,
            illuminationPercent: illumination,
            activityPrediction: activity
        )
    }

    private static func toJulianDate(_ date: Date) -> Double {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        let components = calendar.dateComponents([.year, .month, .day, .hour, .minute, .second], from: date)

        let year = components.year!
        let month = components.month!
        let day = components.day!
        let hour = components.hour!
        let minute = components.minute!
        let second = components.second!

        let dayFraction = (Double(hour) + Double(minute) / 60.0 + Double(second) / 3600.0) / 24.0
        let a = (14 - month) / 12
        let y = year + 4800 - a
        let m = month + 12 * a - 3
        let jdn = day + (153 * m + 2) / 5 + 365 * y + y / 4 - y / 100 + y / 400 - 32045

        return Double(jdn) + dayFraction - 0.5
    }

    private static func getPhaseFromCycleProgress(_ progress: Double) -> MoonPhase {
        switch progress {
        case ..<0.0625: return .newMoon
        case ..<0.1875: return .waxingCrescent
        case ..<0.3125: return .firstQuarter
        case ..<0.4375: return .waxingGibbous
        case ..<0.5625: return .fullMoon
        case ..<0.6875: return .waningGibbous
        case ..<0.8125: return .lastQuarter
        case ..<0.9375: return .waningCrescent
        default: return .newMoon
        }
    }

    private static func calculateIllumination(_ progress: Double) -> Double {
        let angle = progress * 2 * .pi
        return ((1 - cos(angle)) / 2) * 100
    }

    private static func predictWildlifeActivity(_ phase: MoonPhase) -> WildlifeActivityPrediction {
        switch phase {
        case .fullMoon: return .veryHigh
        case .waxingGibbous, .waningGibbous: return .high
        case .firstQuarter, .lastQuarter: return .medium
        case .waxingCrescent, .waningCrescent: return .low
        case .newMoon: return .veryLow
        }
    }

    static func getNextFullMoon(from date: Date = Date()) -> Date {
        let info = calculateMoonPhase(timestamp: date)
        let daysUntilFull: Double
        if info.daysSinceNewMoon < synodicMonth / 2 {
            daysUntilFull = (synodicMonth / 2) - info.daysSinceNewMoon
        } else {
            daysUntilFull = synodicMonth - info.daysSinceNewMoon + (synodicMonth / 2)
        }
        return date.addingTimeInterval(daysUntilFull * 24 * 60 * 60)
    }

    static func getNextNewMoon(from date: Date = Date()) -> Date {
        let info = calculateMoonPhase(timestamp: date)
        return date.addingTimeInterval(info.daysUntilNextNewMoon * 24 * 60 * 60)
    }
}
